### PR TITLE
Prefer immutable collections in the ide subproject

### LINKED ISSUE #23032: Allow configuration of Eclipse build directory
Currently the [eclipse.classpath.defaultOutputDir](https://docs.gradle.org/current/dsl/org.gradle.plugins.ide.eclipse.model.EclipseClasspath.html#org.gradle.plugins.ide.eclipse.model.EclipseClasspath:defaultOutputDir) (The default output directory where Eclipse puts compiled classes.) is pretty useless. Nothing outputs anything there by default and it's ignored by all sourcesets. eg.:
```
eclipse {
  classpath {
    defaultOutputDir = file('build-eclipse')
  }
}
```
will generate the following, when imported in Eclipse Buildship:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="src" output="bin/main" path="src/main/java">
		<attributes>
			<attribute name="gradle_scope" value="main"/>
			<attribute name="gradle_used_by_scope" value="main,test,integrationTest,native-test"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="src" output="bin/main" path="src/main/resources">
		<attributes>
			<attribute name="gradle_scope" value="main"/>
			<attribute name="gradle_used_by_scope" value="main,test,integrationTest,native-test"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-17/"/>
	<classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer"/>
	<classpathentry kind="output" path="build-eclipse"/>
</classpath>
```
Each sourceset will, no matter what, generate something under the `bin/` directory.
What is actually needed is a  defaultBuildDir, that would allow each sourceset to generate an output directory composed of `${defaultBuildDir}/${configuration}`

And no, rewriting .classpath files after the fact is not a serious solution.

### Context
VS Code Java uses BuildShip under the hood. Its users have been complaining about those bin/ directories being created, classes being added to existing bin/ directories with other legit content. See
- https://github.com/redhat-developer/vscode-java/issues/634
- https://github.com/redhat-developer/vscode-java/issues/1615
- https://github.com/redhat-developer/vscode-java/issues/2344

See related Buildship request
- https://github.com/eclipse/buildship/issues/1131

### PR DESCRIPTION
**Pull request by reinsch82**
- refactoring during #23032

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/SourceFolder.java [hunk 1]
```diff
@@ -58,8 +57,8 @@ private List<String> parseNodeListAttribute(Node node, String attributeName) {
     public SourceFolder(String projectRelativePath, String output) {
         super(projectRelativePath);
         this.output = normalizePath(output);
-        this.includes = Lists.newArrayList();
-        this.excludes = Lists.newArrayList();
+        this.includes = ImmutableList.of();
+        this.excludes = ImmutableList.of();
```
Comment (donat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/SourceFolder.java:61]: This is a potential breaking change. Although it is a bad practice, users might modify the result of `getExcludes()` or `getIncludes()`. We can 1) revert or 2) document the change.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/ClasspathFactory.java [hunk 2]
```diff
@@ -40,13 +42,13 @@ public ClasspathFactory(EclipseClasspath classpath, IdeArtifactRegistry ideArtif
     }
 
     public List<ClasspathEntry> createEntries() {
-        List<ClasspathEntry> entries = Lists.newArrayList();
+        ImmutableList.Builder<ClasspathEntry> entries = ImmutableList.builder();
```
Comment (donat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/ClasspathFactory.java:45]: This is most probably a breaking change. Please check it with a sample project with a `eclipse.classpath.file.whenMerged { entries.removeAll { it.path == 'src/main/resources' } }` script. This is something that we actually [recommend](https://discuss.gradle.org/t/buildship-1-0-18-is-now-available/19012).

**Review by donat — CHANGES_REQUESTED** [review]

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/ClasspathFactory.java (hunk 2 above)
Comment (reinsch82) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/ClasspathFactory.java:45]: doesn't this example refer to [this](
https://github.com/reinsch82/gradle/blob/ea05cc46c6697b79e2514b85c0804418fae520ef/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/Classpath.java#L42)?
Which is retrieved by ´getEntries´?
the ´createEntries´shouldn't end up there, or do I miss something?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/SourceFolder.java (hunk 1 above)
Comment (reinsch82) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/SourceFolder.java:61]: I'd revert it in this case.

Comment (reinsch82) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/SourceFolder.java:61]: [these](https://github.com/gradle/gradle/blob/ea05cc46c6697b79e2514b85c0804418fae520ef/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/SourceFolder.java#L53) are also ending up on that variable, which is also not modifiable.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/ClasspathFactory.java (hunk 2 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/ClasspathFactory.java:45]: You are right, I was just overly cautious.

**Review by donat — APPROVED** [review]

**Comment by donat** [comment]
@bot-gradle test and merge

**Pull request merged**